Author: Danny Milosavljevic <dannym@friendly-machines.com>
Date: 2026-06-06
Subject: Prune ogcoff when COFF targets are disabled
License: expat

The tp2cc bootstrap disables COFF target systems with NOTARGET* defines, but
FPC 3.2.0 still lists ogcoff in the CPU target unit `uses` clauses
unconditionally.  That makes the translated bootstrap compile the COFF writer
even though no enabled target can use it.  ogcoff then fails translation on
packed `coffbigobjsymbol.name` accesses and Default(tcoffbigobjheader).

Guard the direct ogcoff imports with the same COFF-target predicates that the
bootstrap already defines.  This also keeps ogcoff's initialization section out
of the translated compiler when all COFF targets are disabled.

--- a/compiler/i386/cputarg.pas
+++ b/compiler/i386/cputarg.pas
@@ -104,6 +104,8 @@ implementation
     {$endif}
 
-      ,ogcoff
+    {$if not defined(NOTARGETGO32V2) or not defined(NOTARGETWIN) or not defined(NOTARGETNATIVENT) or not defined(NOTARGETWDOSX) or not defined(NOTARGETWINCE)}
+      ,ogcoff
+    {$endif}
       ,ogelf
       ,ogmacho
       ,cpuelf
--- a/compiler/x86_64/cputarg.pas
+++ b/compiler/x86_64/cputarg.pas
@@ -78,6 +78,8 @@ implementation
     {$endif}
 
-      ,ogcoff
+    {$ifndef NOTARGETWIN}
+      ,ogcoff
+    {$endif}
       ,ogelf
       ,cpuelf
 
--- a/compiler/arm/cputarg.pas
+++ b/compiler/arm/cputarg.pas
@@ -74,6 +74,8 @@ implementation
     {$endif}
 
-      ,ogcoff
+    {$ifndef NOTARGETWINCE}
+      ,ogcoff
+    {$endif}
       ,ogelf
       ,cpuelf
 
